examples/generative/wgan_gp.py

```diff
@@ -0,0 +1,430 @@
+    x = layers.Dense(1)(x)
+
+    d_model = keras.models.Model(img_input, x, name="discriminator")
+    return d_model
+
+
+d_model = get_discriminator_model()
+d_model.summary()
+
+"""
+## Create the generator
+"""
+
+
+def upsample_block(
+    x,
+    filters,
+    activation,
+    kernel_size=(3, 3),
+    strides=(1, 1),
+    up_size=(2, 2),
+    padding="same",
+    use_bn=False,
+    use_bias=True,
+    use_dropout=False,
+    drop_value=0.3,
+):
+
+    x = layers.UpSampling2D(up_size)(x)
+    x = layers.Conv2D(
+        filters, kernel_size, strides=strides, padding=padding, use_bias=use_bias
+    )(x)
+
+    if use_bn:
+        x = layers.BatchNormalization()(x)
+
+    if activation:
+        x = activation(x)
+    if use_dropout:
+        x = layers.Dropout(drop_value)(x)
+    return x
+
+
+def get_generator_model():
+    noise = layers.Input(shape=(noise_dim,))
+    x = layers.Dense(4 * 4 * 256, use_bias=False)(noise)
+    x = layers.BatchNormalization()(x)
+    x = layers.LeakyReLU(0.2)(x)
+
+    x = layers.Reshape((4, 4, 256))(x)
+    x = upsample_block(
+        x,
+        128,
+        layers.LeakyReLU(0.2),
+        strides=(1, 1),
+        use_bias=False,
+        use_bn=True,
+        padding="same",
+        use_dropout=False,
+    )
+    x = upsample_block(
+        x,
+        64,
+        layers.LeakyReLU(0.2),
+        strides=(1, 1),
+        use_bias=False,
+        use_bn=True,
+        padding="same",
+        use_dropout=False,
+    )
+    x = upsample_block(
+        x, 1, layers.Activation("tanh"), strides=(1, 1), use_bias=False, use_bn=True
+    )
+    # At this point, we have an output which has the same shape as input (32, 32, 1).
+    # We will simply use Cropping2D layer to crop and make it (28, 28, 1).
+    x = layers.Cropping2D((2, 2))(x)
+
+    g_model = keras.models.Model(noise, x, name="generator")
+    return g_model
+
+
+g_model = get_generator_model()
+g_model.summary()
+
+"""
+## Create a WGAN-GP model
+
+Now that we have defined our generator and discriminator models, we will
+implement the WGAN-GP model. We will override the `train_step` for training.
+"""
+
+
+class WGAN(keras.Model):
+    def __init__(
+        self,
+        discriminator,
+        generator,
+        latent_dim,
+        discriminator_extra_steps=3,
+        gp_weight=10.0,
+    ):
+        super(WGAN, self).__init__()
+        self.discriminator = discriminator
+        self.generator = generator
+        self.latent_dim = latent_dim
+        self.d_steps = discriminator_extra_steps
+        self.gp_weight = gp_weight
+
+    def compile(self, d_optimizer, g_optimizer, d_loss_fn, g_loss_fn):
+        super(WGAN, self).compile()
+        self.d_optimizer = d_optimizer
+        self.g_optimizer = g_optimizer
+        self.d_loss_fn = d_loss_fn
+        self.g_loss_fn = g_loss_fn
+
+    def gradient_penalty(self, batch_size, real_images, fake_images):
+        """ The gradient penalty loss is calculated on an interpolated
```

Comment:
The first line in a method docstring should be a standalone sentence that ends with a period. It should fit in one line.

You can add details in the following lines (insert a blank line in between).